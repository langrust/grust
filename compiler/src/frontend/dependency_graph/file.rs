use petgraph::algo::toposort;
use petgraph::graphmap::DiGraphMap;

prelude! {
    error::{Error, TerminationError},
    hir::file::File,
    symbol_table::SymbolTable,

}

impl File {
    /// Generate dependency graph for every nodes/component.
    pub fn generate_dependency_graphs(
        &mut self,
        symbol_table: &SymbolTable,
        errors: &mut Vec<Error>,
    ) -> Result<(), TerminationError> {
        let File { nodes, .. } = self;

        // initialize dictionariy for reduced graphs
        let mut nodes_reduced_graphs = HashMap::new();

        // create graph of nodes
        let mut nodes_graph = DiGraphMap::new();
        nodes
            .iter()
            .for_each(|node| node.add_node_dependencies(&mut nodes_graph));

        // sort nodes according to their dependencies
        let sorted_nodes = toposort(&nodes_graph, None).map_err(|node| {
            let error = Error::NotCausalNode {
                node: symbol_table.get_name(node.node_id()).clone(),
                location: self.location.clone(),
            };
            errors.push(error);
            TerminationError
        })?;
        nodes.sort_by(|n1, n2| {
            let index1 = sorted_nodes
                .iter()
                .position(|id| *id == n1.id)
                .expect("should be in sorted list");
            let index2 = sorted_nodes
                .iter()
                .position(|id| *id == n2.id)
                .expect("should be in sorted list");

            Ord::cmp(&index2, &index1)
        });

        // ordered nodes complete their dependency graphs
        nodes
            .iter_mut()
            .map(|node| node.compute_dependencies(symbol_table, &mut nodes_reduced_graphs, errors))
            .collect::<Vec<Result<(), TerminationError>>>()
            .into_iter()
            .collect::<Result<(), TerminationError>>()?;

        Ok(())
    }
}
