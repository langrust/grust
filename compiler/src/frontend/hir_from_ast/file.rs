prelude! {
    ast::Ast,
    hir::interface::Interface,
}

use super::HIRFromAST;

impl HIRFromAST for Ast {
    type HIR = hir::File;

    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        // initialize symbol table with builtin operators
        symbol_table.initialize();

        // store elements in symbol table
        self.store(symbol_table, errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Function(function) => {
                        functions.push(function.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Typedef(typedef) => {
                        typedefs.push(typedef.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Service(service) => {
                        services.push(service.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Import(import) => imports.push(
                        import
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
