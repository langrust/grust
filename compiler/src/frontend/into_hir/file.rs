prelude! {
    ast::Ast,
    hir::interface::Interface,
}

impl IntoHir<hir::ctx::Simple<'_>> for Ast {
    type Hir = hir::File;

    fn into_hir(self, ctxt: &mut hir::ctx::Simple) -> TRes<Self::Hir> {
        // initialize symbol table with builtin operators
        ctxt.syms.initialize();

        // store elements in symbol table
        self.store(ctxt.syms, ctxt.errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => components.push(component.into_hir(ctxt)),
                    ast::Item::Function(function) => functions.push(function.into_hir(ctxt)),
                    ast::Item::Typedef(typedef) => typedefs.push(typedef.into_hir(ctxt)),
                    ast::Item::Service(service) => services.push(service.into_hir(ctxt)),
                    ast::Item::Import(import) => imports.push(
                        import
                            .into_hir(ctxt)
                            .map(|res| (ctxt.syms.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .into_hir(ctxt)
                            .map(|res| (ctxt.syms.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.into_hir(ctxt))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
