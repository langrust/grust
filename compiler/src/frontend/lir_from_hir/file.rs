use crate::{
    hir::file::File,
    lir::{item::Item, project::Project},
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for File {
    type LIR = Project;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let File {
            typedefs,
            functions,
            nodes,
            interface,
            ..
        } = self;

        let mut typedefs = typedefs
            .into_iter()
            .map(|typedef| typedef.lir_from_hir(symbol_table))
            .collect();
        let mut functions = functions
            .into_iter()
            .map(|function| function.lir_from_hir(symbol_table))
            .map(Item::Function)
            .collect();
        let mut state_machines = nodes
            .into_iter()
            .map(|node| node.lir_from_hir(symbol_table))
            .map(Item::StateMachine)
            .collect();
        let execution_machine = interface.lir_from_hir(symbol_table);

        let mut items = vec![];
        items.append(&mut typedefs);
        items.append(&mut functions);
        items.append(&mut state_machines);
        items.push(Item::ExecutionMachine(execution_machine));

        Project { items }
    }
}
