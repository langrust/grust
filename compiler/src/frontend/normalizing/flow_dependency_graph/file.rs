prelude! { just
    hir::File,
}

impl File {
    /// Generate dependency graphs for the interface.
    #[inline]
    pub fn generate_flows_dependency_graphs(&mut self) {
        self.interface.generate_flows_dependency_graphs()
    }
}
