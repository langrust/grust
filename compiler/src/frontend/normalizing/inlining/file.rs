use std::collections::HashMap;

use crate::{hir::file::File, symbol_table::SymbolTable};

impl File {
    /// Inline node application when it is needed.
    ///
    /// Inlining needed for "shifted causality loop".
    ///
    /// # Example:
    /// ```GR
    /// node semi_fib(i: int) {
    ///     out o: int = 0 fby (i + 1 fby i);
    /// }
    /// node fib_call() {
    ///    out fib: int = semi_fib(fib).o;
    /// }
    /// ```
    /// In this example, `fib_call` calls `semi_fib` with the same input and output signal.
    /// There is no causality loop, `o` depends on the memory of `i`.
    ///
    /// We need to inline the code, the output `fib` is defined before the input `fib`,
    /// which can not be computed by a function call.
    pub fn inline_when_needed(&mut self, symbol_table: &mut SymbolTable) {
        let nodes = self
            .nodes
            .iter()
            .flat_map(|node| {
                node.unitary_nodes
                    .values()
                    .map(|unitary_node| (unitary_node.id.clone(), unitary_node.clone()))
            })
            .collect::<HashMap<_, _>>();
        self.nodes
            .iter_mut()
            .for_each(|node| node.inline_when_needed(&nodes, symbol_table))
    }
}
