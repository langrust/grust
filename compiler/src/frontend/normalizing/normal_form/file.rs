prelude! {
    hir::File,
}

impl File {
    /// Change HIR file into a normal form.
    ///
    /// The normal form of a node is as follows:
    /// - node application can only append at root expression
    /// - node application inputs are signal calls
    /// - no rising edges (replaced by |test| test && ! (false fby test))
    ///
    /// The normal form of a flow expression is as follows:
    /// - flow expressions others than identifiers are root expression
    /// - then, arguments are only identifiers
    ///
    /// # Example
    ///
    /// ```GR
    /// function test(i: int) -> int {
    ///     let x: int = i;
    ///     return x;
    /// }
    /// node my_node(x: int, y: int) {
    ///     out o: int = x*y;
    /// }
    /// node other_node(x: int, y: int) {
    ///     out o: int = x*y;
    /// }
    /// node test(s: int, v: int, g: int) {
    ///     out x: int = 1 + my_node(s, v*2).o;
    ///     out y: int = other_node(g-1, v).o;
    /// }
    /// ```
    ///
    /// The above node contains the following unitary nodes:
    ///
    /// ```GR
    /// node test_x(s: int, v: int) {
    ///     out x: int = 1 + my_node(s, v*2).o;
    /// }
    /// node test_y(v: int, g: int) {
    ///     out y: int = other_node(g-1, v).o;
    /// }
    /// ```
    ///
    /// Which are transformed into:
    ///
    /// ```GR
    /// node test_x(s: int, v: int) {
    ///     x_1: int = v*2;
    ///     x_2: int = my_node(s, x_1).o;
    ///     out x: int = 1 + x_2;
    /// }
    /// node test_y(v: int, g: int) {
    ///     x: int = g-1;
    ///     out y: int = other_node(x_1, v).o;
    /// }
    /// ```
    ///
    /// This example is tested in source.
    pub fn normal_form(&mut self, symbol_table: &mut SymbolTable) {
        let mut nodes_reduced_graphs = HashMap::new();
        // get every nodes' graphs
        self.components.iter().for_each(|node| {
            let test_first_insert =
                nodes_reduced_graphs.insert(node.get_id().clone(), node.get_graph().clone());
            debug_assert!(test_first_insert.is_none())
        });
        // normalize nodes
        self.components
            .iter_mut()
            .for_each(|node| node.normal_form(&nodes_reduced_graphs, symbol_table));

        // normalize interface
        self.interface.normal_form(symbol_table);

        // Debug: test it is in normal form
        debug_assert!(self.is_normal_form());
    }
}
