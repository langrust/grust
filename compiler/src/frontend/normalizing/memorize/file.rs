use crate::hir::file::File;
use crate::symbol_table::SymbolTable;

impl File {
    /// Create memory for HIR nodes' unitary nodes.
    ///
    /// Store buffer for followed by expressions and unitary node applications.
    /// Transform followed by expressions in signal call.
    ///
    /// # Example
    ///
    /// ```GR
    /// node test(s: int, v: int) {
    ///     x_1: int = 0 fby v;
    ///     x_2: int = my_node(s, x_1).o;
    ///     out x: int = 1 + x_2;
    /// }
    /// ```
    ///
    /// The above example becomes:
    ///
    /// ```GR
    /// node test(s: int, v: int) {
    ///     x_1: int = mem;
    ///     x_2: int = my_node(s, x_1).o;
    ///     out x: int = 1 + x_2;
    /// }
    /// memory test {
    ///     buffers: {
    ///         mem: int = 0 fby v;
    ///     },
    ///     called_nodes: {
    ///         memmy_node_o_: (my_node, o);
    ///     },
    /// }
    /// ```
    ///
    /// This example is tested in source.
    pub fn memorize(&mut self, symbol_table: &mut SymbolTable) {
        self.nodes
            .iter_mut()
            .for_each(|node| node.memorize(symbol_table));

        // Debug: test there is no FollowedBy expressions
        debug_assert!(self.no_fby());
    }
}
