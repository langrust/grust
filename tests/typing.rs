use codespan_reporting::files::{Files, SimpleFiles};

use grustine::ast::file::File;
use grustine::error::display;
use grustine::parser::langrust;

#[test]
fn typing_counter() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/counter.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(counter_id, &files.source(counter_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_blinking() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/blinking.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_button_management() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/button_management.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_button_management_condition_match() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_button_management_using_function() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/button_management_using_function.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_pid() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/pid.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(pid_id, &files.source(pid_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn error_when_typing_counter_not_well_typed() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_not_well_typed_id = files.add(
        "counter_not_well_typed.gr",
        std::fs::read_to_string("tests/fixture/counter_not_well_typed.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_not_well_typed_id,
            &files.source(counter_not_well_typed_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_counter_unknown_signal() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_unknown_signal_id = files.add(
        "counter_unknown_signal.gr",
        std::fs::read_to_string("tests/fixture/counter_unknown_signal.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_unknown_signal_id,
            &files.source(counter_unknown_signal_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_unknown_node() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_unknown_node_id = files.add(
        "blinking_unknown_node.gr",
        std::fs::read_to_string("tests/fixture/blinking_unknown_node.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_unknown_node_id,
            &files.source(blinking_unknown_node_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_button_management_using_function_unknown_element() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let button_management_using_function_unknown_element_id = files.add(
        "button_management_using_function_unknown_element.gr",
        std::fs::read_to_string(
            "tests/fixture/button_management_using_function_unknown_element.gr",
        )
        .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            button_management_using_function_unknown_element_id,
            &files
                .source(button_management_using_function_unknown_element_id)
                .unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_button_management_unknown_type() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let button_management_unknown_type_id = files.add(
        "button_management_unknown_type.gr",
        std::fs::read_to_string("tests/fixture/button_management_unknown_type.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            button_management_unknown_type_id,
            &files.source(button_management_unknown_type_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_unknown_field() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_unknown_field_id = files.add(
        "pid_unknown_field.gr",
        std::fs::read_to_string("tests/fixture/pid_unknown_field.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_unknown_field_id,
            &files.source(pid_unknown_field_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_missing_field() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_missing_field_id = files.add(
        "pid_missing_field.gr",
        std::fs::read_to_string("tests/fixture/pid_missing_field.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_missing_field_id,
            &files.source(pid_missing_field_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_component_call() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_component_call_id = files.add(
        "blinking_component_call.gr",
        std::fs::read_to_string("tests/fixture/blinking_component_call.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_component_call_id,
            &files.source(blinking_component_call_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_already_defined_element() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_already_defined_element_id = files.add(
        "blinking_already_defined_element.gr",
        std::fs::read_to_string("tests/fixture/blinking_already_defined_element.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_already_defined_element_id,
            &files.source(blinking_already_defined_element_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_incompatible_type() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_incompatible_type_id = files.add(
        "blinking_incompatible_type.gr",
        std::fs::read_to_string("tests/fixture/blinking_incompatible_type.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_incompatible_type_id,
            &files.source(blinking_incompatible_type_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_incompatible_pattern() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_incompatible_pattern_id = files.add(
        "pid_incompatible_pattern.gr",
        std::fs::read_to_string("tests/fixture/pid_incompatible_pattern.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_incompatible_pattern_id,
            &files.source(pid_incompatible_pattern_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_button_management_using_function_incompatible_input_number() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let button_management_using_function_incompatible_input_number_id = files.add(
        "button_management_using_function_incompatible_input_number.gr",
        std::fs::read_to_string(
            "tests/fixture/button_management_using_function_incompatible_input_number.gr",
        )
        .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            button_management_using_function_incompatible_input_number_id,
            &files
                .source(button_management_using_function_incompatible_input_number_id)
                .unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_counter_expect_number() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_expect_number_id = files.add(
        "counter_expect_number.gr",
        std::fs::read_to_string("tests/fixture/counter_expect_number.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_expect_number_id,
            &files.source(counter_expect_number_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_counter_expect_option() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_expect_option_id = files.add(
        "counter_expect_option.gr",
        std::fs::read_to_string("tests/fixture/counter_expect_option.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_expect_option_id,
            &files.source(counter_expect_option_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_expect_structure_type() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_expect_structure_type_id = files.add(
        "pid_expect_structure_type.gr",
        std::fs::read_to_string("tests/fixture/pid_expect_structure_type.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_expect_structure_type_id,
            &files.source(pid_expect_structure_type_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_unknown_enumeration() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_unknown_enumeration_id = files.add(
        "pid_unknown_enumeration.gr",
        std::fs::read_to_string("tests/fixture/pid_unknown_enumeration.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_unknown_enumeration_id,
            &files.source(pid_unknown_enumeration_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}
