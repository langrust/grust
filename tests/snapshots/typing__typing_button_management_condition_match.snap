---
source: tests/typing.rs
expression: file
---
typedefs:
  - Enumeration:
      id: Button
      elements:
        - Released
        - Pressed
      location: ~
  - Enumeration:
      id: ResetState
      elements:
        - InProgress
        - Confirmed
      location: ~
functions: []
nodes:
  - id: counter
    is_component: false
    inputs:
      - - res
        - Boolean
      - - inc
        - Integer
    equations:
      - - o
        - scope: Output
          id: o
          signal_type: Integer
          expression:
            FunctionApplication:
              function_expression:
                Identifier:
                  id: if_then_else
                  typing:
                    Abstract:
                      - - Boolean
                        - Integer
                        - Integer
                      - Integer
                  location: ~
              inputs:
                - SignalCall:
                    id: res
                    typing: Boolean
                    location: ~
                - Constant:
                    constant:
                      Integer: 0
                    typing: Integer
                    location: ~
                - FunctionApplication:
                    function_expression:
                      Identifier:
                        id: " + "
                        typing:
                          Abstract:
                            - - Integer
                              - Integer
                            - Integer
                        location: ~
                    inputs:
                      - FunctionApplication:
                          function_expression:
                            Identifier:
                              id: (_)
                              typing:
                                Abstract:
                                  - - Integer
                                  - Integer
                              location: ~
                          inputs:
                            - FollowedBy:
                                constant:
                                  Integer: 0
                                expression:
                                  SignalCall:
                                    id: o
                                    typing: Integer
                                    location: ~
                                typing: Integer
                                location: ~
                          typing: Integer
                          location: ~
                      - SignalCall:
                          id: inc
                          typing: Integer
                          location: ~
                    typing: Integer
                    location: ~
              typing: Integer
              location: ~
          location: ~
    contract:
      requires: []
      ensures: []
      invariant: []
      assert: []
    location: ~
  - id: reset_button
    is_component: false
    inputs:
      - - button_state
        - Enumeration: Button
      - - period
        - Integer
      - - reset_limit_1
        - Integer
      - - reset_limit_2
        - Integer
    equations:
      - - user_reset_request_state
        - scope: Output
          id: user_reset_request_state
          signal_type:
            Enumeration: ResetState
          expression:
            Match:
              expression:
                SignalCall:
                  id: button_state
                  typing:
                    Enumeration: Button
                  location: ~
              arms:
                - - Constant:
                      constant:
                        Enumeration:
                          - Button
                          - Released
                      location: ~
                  - ~
                  - Constant:
                      constant:
                        Enumeration:
                          - ResetState
                          - Confirmed
                      typing:
                        Enumeration: ResetState
                      location: ~
                - - Constant:
                      constant:
                        Enumeration:
                          - Button
                          - Pressed
                      location: ~
                  - FunctionApplication:
                      function_expression:
                        Identifier:
                          id: " >= "
                          typing:
                            Abstract:
                              - - Integer
                                - Integer
                              - Boolean
                          location: ~
                      inputs:
                        - SignalCall:
                            id: counter
                            typing: Integer
                            location: ~
                        - SignalCall:
                            id: reset_limit_1
                            typing: Integer
                            location: ~
                      typing: Boolean
                      location: ~
                  - Constant:
                      constant:
                        Enumeration:
                          - ResetState
                          - InProgress
                      typing:
                        Enumeration: ResetState
                      location: ~
                - - Default:
                      location: ~
                  - ~
                  - Constant:
                      constant:
                        Enumeration:
                          - ResetState
                          - Confirmed
                      typing:
                        Enumeration: ResetState
                      location: ~
              typing:
                Enumeration: ResetState
              location: ~
          location: ~
      - - user_reset_request
        - scope: Output
          id: user_reset_request
          signal_type: Boolean
          expression:
            Match:
              expression:
                SignalCall:
                  id: button_state
                  typing:
                    Enumeration: Button
                  location: ~
              arms:
                - - Constant:
                      constant:
                        Enumeration:
                          - Button
                          - Released
                      location: ~
                  - ~
                  - FollowedBy:
                      constant:
                        Boolean: false
                      expression:
                        SignalCall:
                          id: user_reset_request
                          typing: Boolean
                          location: ~
                      typing: Boolean
                      location: ~
                - - Constant:
                      constant:
                        Enumeration:
                          - Button
                          - Pressed
                      location: ~
                  - ~
                  - FunctionApplication:
                      function_expression:
                        Identifier:
                          id: " >= "
                          typing:
                            Abstract:
                              - - Integer
                                - Integer
                              - Boolean
                          location: ~
                      inputs:
                        - SignalCall:
                            id: counter
                            typing: Integer
                            location: ~
                        - FunctionApplication:
                            function_expression:
                              Identifier:
                                id: " + "
                                typing:
                                  Abstract:
                                    - - Integer
                                      - Integer
                                    - Integer
                                location: ~
                            inputs:
                              - SignalCall:
                                  id: reset_limit_1
                                  typing: Integer
                                  location: ~
                              - SignalCall:
                                  id: reset_limit_2
                                  typing: Integer
                                  location: ~
                            typing: Integer
                            location: ~
                      typing: Boolean
                      location: ~
              typing: Boolean
              location: ~
          location: ~
      - - counter
        - scope: Local
          id: counter
          signal_type: Integer
          expression:
            NodeApplication:
              node: counter
              inputs:
                - SignalCall:
                    id: res
                    typing: Boolean
                    location: ~
                - SignalCall:
                    id: period
                    typing: Integer
                    location: ~
              signal: o
              typing: Integer
              location: ~
          location: ~
      - - res
        - scope: Local
          id: res
          signal_type: Boolean
          expression:
            FunctionApplication:
              function_expression:
                Identifier:
                  id: " && "
                  typing:
                    Abstract:
                      - - Boolean
                        - Boolean
                      - Boolean
                  location: ~
              inputs:
                - FunctionApplication:
                    function_expression:
                      Identifier:
                        id: (_)
                        typing:
                          Abstract:
                            - - Boolean
                            - Boolean
                        location: ~
                    inputs:
                      - FunctionApplication:
                          function_expression:
                            Identifier:
                              id: " == "
                              typing:
                                Abstract:
                                  - - Enumeration: Button
                                    - Enumeration: Button
                                  - Boolean
                              location: ~
                          inputs:
                            - SignalCall:
                                id: button_state
                                typing:
                                  Enumeration: Button
                                location: ~
                            - Constant:
                                constant:
                                  Enumeration:
                                    - Button
                                    - Pressed
                                typing:
                                  Enumeration: Button
                                location: ~
                          typing: Boolean
                          location: ~
                    typing: Boolean
                    location: ~
                - FunctionApplication:
                    function_expression:
                      Identifier:
                        id: (_)
                        typing:
                          Abstract:
                            - - Boolean
                            - Boolean
                        location: ~
                    inputs:
                      - FollowedBy:
                          constant:
                            Boolean: true
                          expression:
                            FunctionApplication:
                              function_expression:
                                Identifier:
                                  id: " == "
                                  typing:
                                    Abstract:
                                      - - Enumeration: Button
                                        - Enumeration: Button
                                      - Boolean
                                  location: ~
                              inputs:
                                - SignalCall:
                                    id: button_state
                                    typing:
                                      Enumeration: Button
                                    location: ~
                                - Constant:
                                    constant:
                                      Enumeration:
                                        - Button
                                        - Released
                                    typing:
                                      Enumeration: Button
                                    location: ~
                              typing: Boolean
                              location: ~
                          typing: Boolean
                          location: ~
                    typing: Boolean
                    location: ~
              typing: Boolean
              location: ~
          location: ~
    contract:
      requires: []
      ensures: []
      invariant: []
      assert: []
    location: ~
component: ~
location: ~

